### ISSUE TITLE
Authenticate with private keys

### ISSUE THREAD

**Issue by dholms** [fission-suite/cli#63]
# Summary
We want to switch cli authentication over to private keys. This is more secure than basic auth, keeps us from storing auth details in `.fission.yaml`, and is a good "testing ground" for our ideas around identity since we own the entire flow

## Problem
Authentication is currently done through a BasicAuth strategy w/ username and password.

## Solution
Store private keys locally (in `~/.ssh`) and sign http requests with those. Use a user-friendly key recovery scheme such as ZeroWallet to make `register` and `login` feel like a traditional account

**Comment by dholms** [comment]
### TLDR
Spent some time today investigating different options. ZeroWallet is mostly figured out since I created the Fission Safe proof of concept app. However, we will need to reimplement it in Haskell.

I mainly looked at how authentication will work. I actually think JWTs might be a good way forward for this. They accomplish 90% of what we want. We just need to adapt the model slightly. Traditionally JWTs are issued by the server and use symmetric key encryption. We would essentially flip the model and have users issue their own keys with asymmetric encryption that can then be verified by the server.

---
Here's my full writeup from today:

**store keys in `~/.ssh`**
- _how does this work with windows??_
- _should we force users to create a new keypair or can they use existing?_
- I prefer the former because:
  - if they use existing, they essentially forgo "recovery"
  - also ssh keys are normally RSA. I imagine we want to standardize around ECC

**zerowallet setup**
- `register` prompt for username/password/email
- use username/password to generate the second shard of ZW scheme
  - use just this shard as a private key
  - this is actual similar to `hedgehog.js` but uses a method that's harder for our server to bruteforce (or an attacker that breaks into our db)
  - the reason for 2 passwords is so that you can store just a shard locally and "unlock" it on every transaction with a password. Using the 2nd password to recover the local shard if needed
  - we don't want a user to have enter a password for _every_ interaction with the cli
  - we can reasonably assume that the `.ssh` folder is treated as "secure"
  - when we start attaching more things to identity, we can give this key a "less trusted" permission set that only allows for cli interaction and not data encryption or value transfer
- store server's artifacts in db
- store the privkey in `.ssh`

**login on new computer**
- `login` prompt for username/password
- get server's artifacts
- combine client-side to recover private key and store in `.ssh`

**authentication w/ key**
- Either:
  - signature in msg body / query param
  - signature in http header of some sort
    - overview of Signature v Authorization field: https://github.com/w3c-dvcg/http-signatures/issues/13#issuecomment-416325919
    - although from what I can tell, Signature is normally only used by the server
    - we could create custom headers (ie `x-nonce` and `x-signature`) similar to https://forum.level1techs.com/t/signing-http-requests-to-make-sure-that-the-client-is-not-an-attacker-or-bot/111821
- this will affect using the api normally (ie basic auth)
  - we'd need to implement "tokens" at the same time we implement privkey auth so that heroku apps & apps that use our client lib continue working
- jwt (after looking into it I think this is my preferred option)
  - jwts are normally issued by the server and signed with symmetric key encryption
  - we can "flip" this. Have a user generate a jwt w/ a private key & let the server check it w/ the public key
  - idea from here: https://blog.miguelgrinberg.com/post/json-web-tokens-with-public-key-signatures
  - we can write new instances for Servant that use our encryption scheme: https://hackage.haskell.org/package/servant-auth-server-0.4.4.0/docs/Servant-Auth-Server-Internal-JWT.html
  - _what do we include in the jwt payload?_
    - `username` (minimum)
    - `userID` (if this makes db lookups easier)
    - `expirationTime` (instead of nonce)
      - keep a very tight window (_30? 60? 300? seconds_)
      - without this, anyone that gets ahold of the jwt basically has unlimited access
      - with this, the jwt will change each time
    - `platform`
      - ie: "heroku"
      - this would allow users to create a jwt for a specific platform that allows only a subset of actions (ie uploading files, but not updating dns)
  - when the server decodes
    - check that the current time is not past the expiration date
    - get username from jwt
    - get public key for username from database
    - verify signature on jwt came from public key
    - Authenticated 🎉
  - the downside of this: the signature only proves the authenticity of the jwt payload (ie `username` and `time`)
    - basically proves "I sent _a_ message at this time" instead of "I sent _this_ message at this time"
    - I think this is okay, it's still waayy more secure than basic auth and more secure than OAuth b/c the token is rotated every request instead of every "refresh period"
    - the only attack vector is someone breaks https, grabs the jwt payload and sends fradulent requests during the allotted timeframe

**Comment by expede** [comment]
I'm just going to make some notes as I have time today. Sorry if there's a bit choppy. We'll sync later 👍

From https://jwt.io

> JWTs can be signed using a secret (with the HMAC algorithm) or a public/private key pair using RSA or ECDSA.

Using RSA in JWT is available in the standard

**Comment by expede** [comment]
> what do we include in the jwt payload?
> 
>    username (minimum)
>    userID (if this makes db lookups easier)
>    expirationTime (instead of nonce)
>        keep a very tight window (30? 60? 300? seconds)
>        without this, anyone that gets ahold of the jwt basically has unlimited access
>        with this, the jwt will change each time
>    platform
>        ie: "heroku"
>        this would allow users to create a jwt for a specific platform that allows only a subset of actions (ie uploading files, but not updating dns)

Why do we need all this in the payload? Some non-obvious information:

* The public key is recoverable directly from the message signature
  * We can use the public key as a global user ID
* Wall clock expiration time is less reliable and less powerful than logic clocks (nonce)
* Delegated auth (i.e. temp token) can be replaced by a verifiable claim


Some genuine questions (what am I missing?)
* Why do we care about the platform?
* Why do we care about the username?

**Comment by dholms** [comment]
> The public key is recoverable directly from the message signature 
Yup totally. Forgot about that. So we can leave off `username` and `userID` in this case and, as you said, just use the public key as the main identifier.

> Wall clock expiration time is less reliable and less powerful than logic clocks (nonce)
But we're already allowing a grace period, so we're already relying on wall clock time right?

> Delegated auth (i.e. temp token) can be replaced by a verifiable claim
not quite getting this, explain?

> Why do we care about the platform?
This could be replaced with `permissions` or something similar. My thought was that you don't want heroku to store your keys for instance. So you might create a jwt that heroku uses to sign messages. Your application may only use `get` and `add` functionality and you don't want it to have access to dns updates, (or billing/data/etc once we have it). Maybe this is addressed by the verifiable claim you mentioned?

> Why do we care about the username?
We don't if we use public key as a global userID :+1:

**Comment by bmann** [comment]
I think we want username as separate from public key. Public keys can be used directly, but I think Fission specific unique usernames can be used for convenience (a bit like Keybase). 

Don’t know where / if that should get baked in.

**Comment by dholms** [comment]
yup we'll still have both :+1: 
That's just what gets sent along in the jwt payload to identify the user by.

Right now the database has `userID` & `username`
The plan would be to transition that over to `pubKey` & `username`

**Comment by expede** [comment]
UPDATE
This is in progress based on an informal spec. Formal spec coming in the whitepaper.
